plenum/test/test_stashing_router.py

```diff
@@ -0,0 +1,193 @@
+    bus.send(some_message)
+    some_handler.assert_called_once_with(some_message)
+    other_handler.assert_not_called()
+
+    other_message = create_other_message()
+    bus.send(other_message)
+    some_handler.assert_called_once_with(some_message)
+    other_handler.assert_called_once_with(other_message)
+
+
+def test_stashing_router_correctly_handles_multiple_arguments():
+    handler = Mock()
+
+    bus = InternalBus()
+    router = StashingRouter(10)
+    router.subscribe(SomeMessage, handler)
+    router.subscribe_to(bus)
+
+    message = create_some_message()
+    bus.send(message, 'hello')
+    handler.assert_called_once_with(message, 'hello')
+
+
+def test_stashing_router_can_stash_messages():
+    stash_count = 3
+    calls = []
+
+    def handler(msg):
+        nonlocal stash_count
+        calls.append(msg)
+        if stash_count > 0:
+            stash_count -= 1
+            return STASH
+
+    bus = InternalBus()
+    router = StashingRouter(10)
+    router.subscribe(SomeMessage, handler)
+    router.subscribe_to(bus)
+
+    msg_a = create_some_message()
+    msg_b = create_some_message()
+    bus.send(msg_a)
+    bus.send(msg_b)
+    assert router.stash_size() == 2
+    assert calls == [msg_a, msg_b]
+
+    router.process_stashed()
+    assert router.stash_size() == 1
+    assert calls == [msg_a, msg_b, msg_a, msg_b]
+
+    router.process_stashed()
+    assert router.stash_size() == 0
+    assert calls == [msg_a, msg_b, msg_a, msg_b, msg_a]
+
+    router.process_stashed()
+    assert router.stash_size() == 0
+    assert calls == [msg_a, msg_b, msg_a, msg_b, msg_a]
+
+
+def test_stashing_router_can_stash_messages_with_different_reasons():
+    calls = []
+
+    def handler(message: SomeMessage):
+        calls.append(message)
+        if message.int_field % 2 == 0:
+            return STASH + 0
+        else:
+            return STASH + 1
+
+    bus = InternalBus()
+    router = StashingRouter(10)
+    router.subscribe(SomeMessage, handler)
+    router.subscribe_to(bus)
+
+    messages = [create_some_message() for _ in range(10)]
+    for msg in messages:
+        bus.send(msg)
+    assert router.stash_size() == len(messages)
+    assert router.stash_size(STASH + 0) + router.stash_size(STASH + 1) == router.stash_size()
+
+    calls.clear()
+    router.process_stashed()
+    assert router.stash_size() == len(messages)
+    assert calls == sorted(messages, key=lambda m: m.int_field % 2)
+
+    calls.clear()
+    router.process_stashed(STASH + 0)
+    assert router.stash_size() == len(messages)
+    assert router.stash_size(STASH + 0) == len(calls)
+    assert all(msg.int_field % 2 == 0 for msg in calls)
+    assert all(msg in messages for msg in calls)
+
+    calls.clear()
+    router.process_stashed(STASH + 1)
+    assert router.stash_size() == len(messages)
+    assert router.stash_size(STASH + 1) == len(calls)
+    assert all(msg.int_field % 2 != 0 for msg in calls)
+    assert all(msg in messages for msg in calls)
+
```

Comment:
Please add a test when `process_stashed` is called (both with and without explicit code arg) without any messages stashed